doc/generated/variables.gen

```diff
@@ -4767,11 +4767,488 @@ A non-default tool list is specified that does not contain any of the msvc tools
 </itemizedlist>
 </para>
 
+<para>
+Important usage details:
+<itemizedlist>
+
+<listitem><para>
+<envar>MSVC_NOTFOUND_POLICY</envar> must be passed as an argument to the Environment() constructor;
+setting it later has no effect.
```

Comment:
Actually just needs to be set before msvc is initialized. Which could be later if empty or non-default tool list is provided.

Reply:
I'm confused.  Ignoring the default environment, MSVC_NOTFOUND_POLICY is the "per-environment" construction variable.  It would need to be set prior to msvc_setup_env_once being called which kicks of autodetection.  I didnt' think there was time between the Environment constructor being called and the msvc_setup_env_once call to make changes to the environment.

Reply:
Yup:  when there are no detected versions of msvc and a non-default tool  is detected.

If at least one version of msvc is detected, I believe it would have to be set in the constructor though.  Correct?

Reply:
Yes.  But note.. a non-trivial percent of users do explicitly list tools and/or start empty and add one by one..

Reply:
I'll revert the MSVC_NOTFOUND_POLICY documentation changes in the morning.

>Yes. But note.. a non-trivial percent of users do explicitly list tools and/or start empty and add one by one..

For the default environment, a specific msvc version, or both?

Reply:
I wouldn't revert. Probably worth just adding, if you're defaulting the tools or loading msvc when you create the environment it must be set then, otherwise it must be set before you load msvc tool.

Reply:
@bdbaddog Questions for you!

It appears that the current main documentation is inaccurate in two places as well.  That's what I get for copy-and-pasting from ```MSVC_UWP_APP``` since I was changing the implementation.

Extending your example:

```
# Current main documentation
#
#   MSVC_UWP_APP
#     This variable must be passed as an argument to the Environment() constructor; setting it later has no effect.
#
#   MSVC_VERSION
#     This variable must be passed as an argument to the Environment constructor; setting it later has no effect.

DefaultEnvironment(tools=[])
env=Environment(tools=[])
env['MSVC_NOTFOUND_POLICY']='GiantMonkey' # my favorite policy ;)
env['MSVC_VERSION']='14.1'
env['MSVC_UWP_APP']='1'
env.Tool('msvc')
```
Both ```MSVC_VERSION``` and ```MSVC_UWP_APP``` can be set after the constructor is called and still have an effect.

**Question 1: Is the following correct?**

*ENVAR_NAME_HERE must be passed as an argument to the environment constructor when an msvc tool (e.g., 'msvc', 'mslink', etc.) is loaded via the default tools list or via a tools list passed to the environment constructor. Otherwise, ENVAR_NAME_HERE must be set before the first msvc tool is loaded into the environment.*

**Question 2: Should ```MSVC_VERSION``` be updated as well?**

Existing documentation fragment:
```
MSVC_USE_SCRIPT
    ...
    MSVC_USE_SCRIPT overrides MSVC_VERSION and TARGET_ARCH
```
To be precise/pedantic, ```MSVC_USE_SCRIPT``` ***ignores*** ```MSVC_VERSION``` and ```TARGET_ARCH``` and would leave them with erroneous values if they were not correct for the resulting script output.  

It also ignores ```HOST_ARCH``` if that were different from the actual host (e.g., 'x86' on 'amd64').

Someone might infer that ```MSVC_VERSION``` and ```TARGET_ARCH``` were being ***overwritten*** based on the results of running the script  which is not the case.

```
MSVC_USE_SCRIPT ignores MSVC_VERSION and TARGET_ARCH and does not update these values after script execution.
```

**Question 3: Any changes necessary?**
I'm ok with not changing anything.  I'm likely reading too much into the original text.

Reply:
Question 1: Yes looks good
Question 2: Yes

`HOST_ARCH` should be read only and set by SCons if I remember correctly, though I guess x64 and x86 would be both be reasonable for modern x86 cpus.

Hmm re `MSVC_VERSION` and `TARGET_ARCH`, maybe better to say they're ignored.
Though there is an existing bug to have MSVC_VERSION be set to the version which is actually used (especially in the case where it wasn't specified).

I'll do another pass on the docs and such (hopefully) later today.

Reply:
> Though there is an existing bug to have MSVC_VERSION be set to the version which is actually used (especially in the case where it wasn't specified).

There is a 1:1 mapping from cl.exe version to toolset version (VS2015 and earlier, this is the MSVC_VERSION).  For VS2017 and later, there is an VS version environment variable that can be mapped to the VS product (MSVC_VERSION).  I'm doing this in my own work.  The only real requirement is running cl.exe which can be done from the same call as setting up the environment subject to some "minor changes".

Reply:
Right.  
So in theory going forward for 2017 -> there'd be MSVC_VERSION and MSVC_TOOLSET_VERSION, which should be set if the user defaulted either or both of these when requesting MSVC tools?
And prior to 2017, the single variable would suffice?

Reply:
> So in theory going forward for 2017 -> there'd be MSVC_VERSION and MSVC_TOOLSET_VERSION, which should be set if the user defaulted either or both of these when requesting MSVC tools?
And prior to 2017, the single variable would suffice?

Pretty much.

One issue is that for input the toolset can be a full specification ('14.29.12345').  However, cl.exe is only going to provide enough information to derive a "major" toolset version (14.2) which could be different from the product ('14.3').  I'm not sure setting it as a toolset is desirable unless variable formats are acceptable.

It may be easiest to have a "synthetic" toolset version for VS2015 and earlier (e.g., '14.00.00000', '12.00.00000', '11.00.00000', etc) which is trivial.  This way all versions have both a MSVC_VERSION and MSVC_TOOLSET_VERSION for user queries and selection.

What may need to be looked at is *how* the variables are being used in some of the tools.  For example, in msvc.py:
```
    if env.get('MSVC_VERSION',False):
        maj, min = msvc_version_to_maj_min(env['MSVC_VERSION'])
        if maj < 8:
            env['PCHPDBFLAGS'] = SCons.Util.CLVar(['${(PDB and "/Yd") or ""}'])
        else:
            env['PCHPDBFLAGS'] = ''
```

At some point does the "backend" need to switch?   Might it matter as the "major" toolset prefix (i.e., 'XX.Y') gets further away from the MSVC_VERSION: currently the widest gap is '14.3' (MSVC_VERSION) and '14.0' (MSVC_TOOLSET_VERSION).

The backend code probably is better thought of as using the major toolset version rather than product version.

Another way to go is to slowly move away from MSVC_VERSION as a user input and make it a derived value as additional desirable enhancements are made (e.g., preview versions, toolset selection, etc). 

MSVC_VERSION could be gradually be replaced with MSVS_PRODUCT (visual studio year), MSVS_CHANNEL (release vs preview) and MSVS_COMPONENT (enterprise, professional, community, express) which are all properties of the Visual Studio "container".

MSVC_TOOLSET_VERSION and MSVC_BUILDTOOLS_VERSION could be read and derived from the visual studio filesystem.

MSVC_VERSION *can* be used alongside the new variables and  *could* be eventually deprecated.